hawkular-inventory-api/src/main/java/org/hawkular/inventory/api/model/AbstractHashTree.java

```diff
@@ -0,0 +1,283 @@
+        T construct(RelativePath path, H hash, Map<Path.Segment, T> children);
+    }
+
+    /**
+     * An interface that is implemented by all tree hash builders. This is only made public for the cases where the
+     * users would like to abstract over different tree implementations.
+     *
+     * @param <This> the type of the concrete implementation
+     * @param <Child> the type of the tree child builder
+     * @param <Tree> the type of the tree being built
+     * @param <Hash> the type of the hash used in the tree
+     */
+    public interface Builder<This extends Builder<This, Child, Tree, Hash>,
+            Child extends ChildBuilder<Child, This, ?, Tree, Hash>,
+            Tree extends AbstractHashTree<Tree, Hash>, Hash extends Serializable> {
+
+        This withPath(RelativePath path);
+
+        This withHash(Hash hash);
+
+        Hash getHash();
+
+        RelativePath getPath();
+
+        boolean hasChildren();
+
+        Child startChild();
+
+        void addChild(Tree childTree);
+    }
+
+    /**
+     * Interface for top-level hash tree builders. I.e. only these actually have the build method that is used to
+     * produce the final tree.
+     *
+     * @param <This> the type of the concrete implementation
+     * @param <Child> the type of the tree child builder
+     * @param <Tree> the type of the tree being built
+     * @param <Hash> the type of the hash used in the tree
+     */
+    public interface TopBuilder<This extends TopBuilder<This, Child, Tree, Hash>,
+            Child extends ChildBuilder<Child, This, ?, Tree, Hash>,
+            Tree extends AbstractHashTree<Tree, Hash>, Hash extends Serializable>
+            extends Builder<This, Child, Tree, Hash> {
+
+        Tree build();
+    }
+
+    /**
+     * A builder used to build sub trees of some parent tree node.
+     *
+     * @param <This> the type of the concrete implementation of this interface
+     * @param <Parent> the type of the builder of the parent tree node
+     * @param <Child> the type of the builder used to build the subtrees of the tree node built by this builder
+     * @param <Tree> the type of the tree being built
+     * @param <Hash> the type of the hash used in the tree
+     */
+    public interface ChildBuilder<This extends ChildBuilder<This, Parent, Child, Tree, Hash>,
+            Parent extends Builder<?, This, Tree, Hash>,
+            Child extends ChildBuilder<Child, This, ?, Tree, Hash>,
+            Tree extends AbstractHashTree<Tree, Hash>,
+            Hash extends Serializable>
+
+            extends Builder<This, Child, Tree, Hash> {
+
+
+        Parent getParent();
+
+        Parent endChild();
+    }
+
+    static class AbstractBuilder<This extends Builder<This, Child, T, H>,
+            Child extends ChildBuilder<Child, This, ?, T, H>,
+            T extends AbstractHashTree<T, H>, H extends Serializable>
+            implements Builder<This, Child, T, H> {
```

Comment:
It is. I wanted to be DRY as much as possible, but maybe something like this could also be achieved with overriden methods with specialized return types at the expense of more code.